source/channel.c

```diff
@@ -805,27 +814,57 @@ struct aws_io_message *aws_channel_slot_acquire_max_message_for_write(struct aws
     return aws_channel_acquire_message_from_pool(slot->channel, AWS_IO_MESSAGE_APPLICATION_DATA, size_hint);
 }
 
-int aws_channel_slot_increment_read_window(struct aws_channel_slot *slot, size_t window) {
+static void s_window_update_task(struct aws_channel_task *channel_task, void *arg, enum aws_task_status status) {
+    (void)channel_task;
+    struct aws_channel *channel = arg;
 
-    if (slot->channel->channel_state < AWS_CHANNEL_SHUTTING_DOWN) {
-        size_t temp = slot->window_size + window;
-        if (temp < slot->window_size) {
-            slot->window_size = SIZE_MAX;
-        } else {
-            slot->window_size = temp;
+    if (status == AWS_TASK_STATUS_RUN_READY && channel->channel_state < AWS_CHANNEL_SHUTTING_DOWN) {
+        /* get the right-most slot to start the updates. */
+        struct aws_channel_slot *slot = channel->first;
+        for (;;) {
+            if (slot->adj_right) {
+                slot = slot->adj_right;
+            } else {
+                break;
+            }
         }
 
-        if (slot->adj_left && slot->adj_left->handler) {
-            AWS_LOGF_TRACE(
-                AWS_LS_IO_CHANNEL,
-                "id=%p: sending increment read window of size %llu, "
-                "on slot %p and notifying slot %p with handler %p.",
-                (void *)slot->channel,
-                (unsigned long long)window,
-                (void *)slot,
-                (void *)slot->adj_left,
-                (void *)slot->adj_left->handler);
-            return aws_channel_handler_increment_read_window(slot->adj_left->handler, slot->adj_left, window);
+        while (slot->adj_left) {
+            struct aws_channel_slot *upstream_slot = slot->adj_left;
+            if (upstream_slot->handler) {
+                slot->window_size = aws_add_size_saturating(slot->window_size, slot->current_window_update_batch_size);
+                size_t update_size = slot->current_window_update_batch_size;
+                slot->current_window_update_batch_size = 0;
+                if (aws_channel_handler_increment_read_window(upstream_slot->handler, upstream_slot, update_size)) {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_IO_CHANNEL,
+                        "channel %p: channel update task failed with status %d",
+                        (void *)slot->channel,
+                        aws_last_error());
+                    slot->channel->window_update_in_progress = false;
+                    aws_channel_shutdown(channel, aws_last_error());
+                    return;
+                }
+
+                slot = slot->adj_left;
+            }
```

Comment:
**IMPORTANT**
`slot = slot->adj_left;` needs to be moved outside the `if (upstream_slot->handler) {` block or we've got an infinite loop

Reply:
doh! fixed. good catch.